src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil/WebcilReader.cs

```diff
@@ -0,0 +1,332 @@
+
+
+public sealed class WebcilReader : IDisposable
+{
+    // WISH:
+    // This should be implemented in terms of System.Reflection.Internal.MemoryBlockProvider like the PEReader,
+    // but the memory block classes are internal to S.R.M.
+
+    private readonly Stream _stream;
+    private WebcilHeader _header;
+    private DirectoryEntry _corHeaderMetadataDirectory;
+    private MetadataReaderProvider? _metadataReaderProvider;
+    private ImmutableArray<WebcilSectionHeader>? _sections;
+
+    public WebcilReader(Stream stream)
+    {
+        this._stream = stream;
+        if (!stream.CanRead || !stream.CanSeek)
+        {
+            throw new ArgumentException("Stream must be readable and seekable", nameof(stream));
+        }
+        if (!ReadHeader())
+        {
+            throw new BadImageFormatException("Stream does not contain a valid Webcil file", nameof(stream));
+        }
+        if (!ReadCorHeader())
+        {
+            throw new BadImageFormatException("Stream does not contain a valid COR header in the Webcil file", nameof(stream));
+        }
+    }
+
+    private unsafe bool ReadHeader()
+    {
+        WebcilHeader header;
+        var buffer = new byte[Marshal.SizeOf<WebcilHeader>()];
+        if (_stream.Read(buffer, 0, buffer.Length) != buffer.Length)
+        {
+            return false;
+        }
+        fixed (byte* p = buffer)
+        {
+            header = *(WebcilHeader*)p;
+        }
+        if (header.id[0] != 'W' || header.id[1] != 'C' || header.version != Internal.Constants.WC_VERSION)
+        {
+            return false;
+        }
+        _header = header;
+        if (!BitConverter.IsLittleEndian)
+        {
+            throw new NotImplementedException("TODO: implement big endian support");
+        }
+        return true;
+    }
+
+    private unsafe bool ReadCorHeader()
+    {
+        // we can't construct CorHeader because it's constructor is internal
+        // but we don't care, really, we only want the metadata directory entry
+        var pos = TranslateRVA(_header.pe_cli_header_rva);
+        if (_stream.Seek(pos, SeekOrigin.Begin) != pos)
+        {
+            return false;
+        }
+        using var reader = new BinaryReader(_stream, System.Text.Encoding.UTF8, leaveOpen: true);
+        reader.ReadInt32(); // byte count
+        reader.ReadUInt16(); // major version
+        reader.ReadUInt16(); // minor version
+        _corHeaderMetadataDirectory = new DirectoryEntry(reader.ReadInt32(), reader.ReadInt32());
+        return true;
+    }
+
+    public MetadataReaderProvider GetMetadataReaderProvider()
+    {
+        // FIXME threading
+        if (_metadataReaderProvider == null)
+        {
+            long pos = TranslateRVA((uint)_corHeaderMetadataDirectory.RelativeVirtualAddress);
+            if (_stream.Seek(pos, SeekOrigin.Begin) != pos)
+            {
+                throw new BadImageFormatException("Could not seek to metadata", nameof(_stream));
```

Comment:
Is `nameof(_stream)` useful here? The caller might have the relevant filename.

Reply:
`nameof(_stream)` is not useful.  But it turns out we don't have any name we can show at all - the DebugStore calls the WebcilReader and it only has an array of bytes from the runtime - until it tries to read the metadata it has no idea what the name of the assembly is.  I'll fix this not to use `nameof` here, but turns out we don't have anything we can show instead.